### PR TITLE
Add geometry shader to mesh shader sample

### PR DESCRIPTION
**Pull request by Dastin-Wilski**
## Description

Sample demonstrates how to use mesh shader to achieve the same results as with geometry shader, it loads model from file and visualizes its normals.

## Notes
Tested on:

Windows 11
AMD Ryzen 5800X
Nvidia RTX 3070 Ti
32 GB RAM

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

### CODE REVIEW

File: framework/api_vulkan_sample.h [hunk 1]
```diff
@@ -217,6 +238,13 @@ class ApiVulkanSample : public vkb::VulkanSample
 	 */
 	std::unique_ptr<vkb::sg::SubMesh> load_model(const std::string &file, uint32_t index = 0);
 
+	/**
+	 * @brief Loads in a single model from a GLTF file to storage buffer using memory alligned vertex strucuture
+	 * @param file The filename of the model to load
+	 * @param index The index of the model to load from the GLTF file (default: 0)
+	 */
+	std::unique_ptr<vkb::sg::SubMesh> load_model_to_storage_buffer(const std::string &file, uint32_t index = 0);
```
Comment (gpx1000) [inline framework/api_vulkan_sample.h]: How about we combine this function into the load_model function.  Give it a boolean variable to say how to store it.  Default to false so existing code should still work.

File: framework/gltf_loader.cpp [hunk 2]
```diff
@@ -442,6 +442,48 @@ std::unique_ptr<sg::SubMesh> GLTFLoader::read_model_from_file(const std::string
 	return std::move(load_model(index));
 }
 
+std::unique_ptr<sg::SubMesh> GLTFLoader::read_model_from_file_to_storage_buffer(const std::string &file_name, uint32_t index)
```
Comment (gpx1000) [inline framework/gltf_loader.cpp]: This also can benefit from reusing the same function, just add a boolean to denote where to save the output (vertex vs storage_buffer).  Should make it easier to maintain.

File: framework/api_vulkan_sample.h (hunk 1 above)
Comment (Dastin-Wilski) [reply framework/api_vulkan_sample.h]: Done.

File: framework/gltf_loader.cpp (hunk 2 above)
Comment (Dastin-Wilski) [reply framework/gltf_loader.cpp]: Done.

File: framework/api_vulkan_sample.h [hunk 3]
```diff
@@ -73,6 +73,27 @@ struct Vertex
 	glm::vec4 weight0;
 };
 
+/**
+ * @brief The structure of a vertex for storage buffer
+ * Simplified to position and normal for easier allignment
+ */
+struct AllignedVertex
```
Comment (asuessenbach) [inline framework/api_vulkan_sample.h]: `AllignedVertex` -> `AlignedVertex`

File: framework/gltf_loader.cpp [hunk 4]
```diff
@@ -1181,6 +1184,182 @@ std::unique_ptr<sg::SubMesh> GLTFLoader::load_model(uint32_t index)
...
+		accessor    = model.accessors[gltf_primitive.attributes.find("NORMAL")->second];
+		buffer_view = model.bufferViews[accessor.bufferView];
+		normals     = reinterpret_cast<const float *>(&(model.buffers[buffer_view.buffer].data[accessor.byteOffset + buffer_view.byteOffset]));
+	}
+
+	for (size_t v = 0; v < vertex_count; v++)
+	{
+		AllignedVertex vert{};
+		vert.pos        = glm::vec4(glm::make_vec3(&pos[v * 3]), 1.0f);
+		vert.normal.xyz = glm::normalize(glm::vec3(normals ? glm::make_vec3(&normals[v * 3]) : glm::vec4(0.0f)));
```
Comment (asuessenbach) [inline framework/gltf_loader.cpp]: Assigning a vec4 onto normal.xyz?
Maybe better:
`vert.normal = normals ? glm::vec4(glm::normalize(glm::make_vec3(&normals[v * 3])), 0.0f) : glm::vec4(0.0f);
`

File: framework/gltf_loader.cpp [hunk 5]
```diff
@@ -1181,6 +1184,182 @@ std::unique_ptr<sg::SubMesh> GLTFLoader::load_model(uint32_t index)
 	return std::move(submesh);
 }
 
+std::unique_ptr<sg::SubMesh> GLTFLoader::load_model_to_storage_buffer(uint32_t index)
```
Comment (asuessenbach) [inline framework/gltf_loader.cpp]: Substantial parts of `load_model` and `load_model_to_storage_buffer` are identical. Please either unify those two functions, or introduce some helper functions that are called from both.

File: framework/gltf_loader.cpp [hunk 6]
```diff
@@ -1181,6 +1184,182 @@ std::unique_ptr<sg::SubMesh> GLTFLoader::load_model(uint32_t index)
...
+		{
+			// index_data is unsigned char type, casting to uint32_t* will give proper value
+			meshlet.indices[meshlet.index_count] = *(reinterpret_cast<uint32_t *>(index_data.data()) + i);
+
+			size_t size1 = vertices.size();
+			vertices.insert(meshlet.indices[meshlet.index_count]);
+			size_t size2 = vertices.size();
+
+			if (size2 > size1)
+				meshlet.vertex_count++;
```
Comment (asuessenbach) [inline framework/gltf_loader.cpp]: Maybe:
```
if (vertices.insert(meshlet.indices[meshlet.index_count]).second)
{
    ++meshlet.vertex_count;
}
```

File: framework/gltf_loader.cpp [hunk 7]
```diff
@@ -1181,6 +1184,182 @@ std::unique_ptr<sg::SubMesh> GLTFLoader::load_model(uint32_t index)
...
+
+			meshlet.index_count++;
+			triangle_check = triangle_check < 3 ? ++triangle_check : 1;
+
+			// 96 because for each traingle we draw a line in a mesh shader sample, 32 triangles/lines per meshlet = 64 vertices on output
+			if (meshlet.vertex_count == 64 || meshlet.index_count == 96)
+			{
+				std::set<uint32_t>::iterator it;
+				uint32_t                     counter = 0;
+				for (it = vertices.begin(); it != vertices.end(); ++it)
```
Comment (asuessenbach) [inline framework/gltf_loader.cpp]: No need to have it available outside this loop.
Maybe:
```
for (auto v : vertices)
{
	meshlet.vertices[counter++] = v;
}
```

File: framework/gltf_loader.cpp [hunk 8]
```diff
@@ -1181,6 +1184,182 @@ std::unique_ptr<sg::SubMesh> GLTFLoader::load_model(uint32_t index)
...
+			default:
+			{
+				break;
+			}
+		}
+
+		// Always do uint32
+		submesh->index_type = VK_INDEX_TYPE_UINT32;
+
+		// prepare meshlets
```
Comment (asuessenbach) [inline framework/gltf_loader.cpp]: Maybe move the meshlets construction into a little helper function.

File: samples/README.md [hunk 9]
```diff
@@ -299,6 +299,11 @@ Demonstrate how to use fragment shader barycentric feature, which allows accessi
 **Extension**: [```VK_EXT_descriptor_buffer```](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VK_ext_descriptor_buffer)
 Demonstrate how to use the new extension to replace descriptor sets with resource descriptor buffers
 
+### [Geometry shader to mesh shader](./extensions/gshader_to_mshader)
+**Extension:** [```VK_EXT_mesh_shader```](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_mesh_shader.html)<br/>
+Demonstrates how to use mesh shader to achieve the same results as with geometry shader, it loads model from file and visualizes its normals.
```
Comment (asuessenbach) [inline samples/README.md]: This is not the only way how you can use a mesh shader as a geometry shader replacement. That is, maybe better: "Demonstrates how a mesh shader can be used to achieve..."

File: samples/extensions/gshader_to_mshader/README.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+-
+-->
+
+# Geometry shader to mesh shader
+
+![Sample](./images/visualization_of_normals.png)
+
+## Overview
+
+This sample demonstrates how to use mesh shader instead of geometry shader to achieve same results. It contains geometry and mesh shader pipelines visualizing normals in teapot model.
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/README.md]: ... visualizing normals in **a** teapot model.
Or maybe even "... in **the** teapot model."

File: samples/extensions/gshader_to_mshader/README.md [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+## Meshlets
+
+To access model vertices from within mesh shader it needs to be stored within storage buffer (SSBO). Indices need to be divided into meshlets and also stored within SSBO so each work item can work on single meshlet.
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/README.md]: storage buffer -> **S**hader **S**torage **B**uffer **O**bject

What is a "meshlet" ?

stored within **an** SSBO ... work on **a** single meshlet.

File: samples/extensions/gshader_to_mshader/README.md [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+This sample adds fuction to framwerok `load_model_to_storage_buffer()` that is based on `load_model()`. It reads model from file, stores vertices positions and normals to SSBO using `VertexAlligned` strcuture (because of std430 memory layout), after that indices are divided into mehslets based on `Meshlet` structure and stored to SSBO.
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/README.md]: A couple of typos and glitches here, maybe instead:
This sample adds the function `load_model_to_storage_buffer()` to the framework which is based on `load_models()`.
It reads the model data from the file `teapot.gltf`, stores vertex positions and normals in an SSBO using the `AlignedVertex` structure (...).
After that, indices are divided into meshlets using the `Meshlet` structure and also stored in SSBOs.

File: samples/extensions/gshader_to_mshader/README.md [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+## Enabling the Extension
+
+To enable mesh shader VulkanAPI 1.1 is required.
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/README.md]: To be precise: 
The device extension is provided by `VK_EXT_MESH_SHADER_EXTENSION_NAME`. It requires `VK_KHR_SPIRV_1_4_EXTENSION_NAME`, which in turn require VulkanAPI 1.1 and `VK_KHR_SHADER_FLOAT_CONTROLS_EXTENSION_NAME`.

File: samples/extensions/gshader_to_mshader/README.md [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+The device extension is provided by `VK_EXT_MESH_SHADER_EXTENSION_NAME`.
+It also requires `VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME`.
+SPIRV needs to be set to 1.4.
+
+```C++
+set_api_version(VK_API_VERSION_1_1);
+add_instance_extension(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/README.md]: `VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME` is not needed here, as it has been promoted to Vulkan 1.1.

File: samples/extensions/gshader_to_mshader/README.md [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+add_device_extension(VK_EXT_MESH_SHADER_EXTENSION_NAME);
+add_device_extension(VK_KHR_SPIRV_1_4_EXTENSION_NAME);
+add_device_extension(VK_KHR_SHADER_FLOAT_CONTROLS_EXTENSION_NAME);
+vkb::GLSLCompiler::set_target_environment(glslang::EShTargetSpv, glslang::EShTargetSpv_1_4);
+```
+The `VkPhysicalDeviceMeshShaderFeaturesEXT` structure needs to be included in the pNext chain of the `VkPhysicalDeviceFeatures2` structure passed to vkGetPhysicalDeviceFeatures2
+
+```C++
+auto &requested_vertex_input_features      = gpu.request_extension_features<VkPhysicalDeviceMeshShaderFeaturesEXT>(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_MESH_SHADER_FEATURES_NV);
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/README.md]: `VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_MESH_SHADER_FEATURES_NV` -> `VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_MESH_SHADER_FEATURES_EXT`

File: samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+ */
+
+#include "gshader_to_mshader.h"
+
+GshaderToMshader::GshaderToMshader()
+{
+	title = "task_mesh_migration";
+
+	set_api_version(VK_API_VERSION_1_1);
+	add_instance_extension(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp]: No need to ask for `VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME`, as it has been promoted to Vulkan 1.1.

File: samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp [hunk 17]
```diff
@@ -0,0 +1,436 @@
...
+		uniform_buffer_ms.reset();
+		object.reset();
+		storage_buffer_object.reset();
+
+		vkDestroyPipeline(get_device().get_handle(), model_pipeline, nullptr);
+		vkDestroyPipeline(get_device().get_handle(), normal_pipeline, nullptr);
+		vkDestroyPipeline(get_device().get_handle(), mesh_pipeline, nullptr);
+		vkDestroyPipelineLayout(get_device().get_handle(), pipeline_layout, nullptr);
+		vkDestroyDescriptorSetLayout(get_device().get_handle(), descriptor_set_layout, nullptr);
+		vkDestroyDescriptorPool(get_device().get_handle(), descriptor_pool, VK_NULL_HANDLE);
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp]: Don't destroy the `descriptor_pool` here, it gets destroyed in the `ApiVulkanSample` destructor!

File: samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp [hunk 18]
```diff
@@ -0,0 +1,436 @@
...
+ * 	@fn void VertexDynamicState::load_assets()
+ *	@brief Loading extra models, textures from assets
+ */
+void GshaderToMshader::load_assets()
+{
+	/* Models */
+
+	object = load_model("scenes/teapot.gltf");
+
+	storage_buffer_object = load_model_to_storage_buffer("scenes/teapot.gltf");
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp]: `load_model_to_storage_buffer` doesn't exist any more! Use `load_model`, instead.

File: samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp [hunk 19]
```diff
@@ -0,0 +1,436 @@
...
+		ubo.model = glm::rotate(ubo.model, glm::pi<float>(), glm::vec3(0, 0, 1));
+	}
+	uniform_buffer_vs->convert_and_update(ubos[0]);
+	uniform_buffer_gs->convert_and_update(ubos[1]);
+	uniform_buffer_ms->convert_and_update(ubos[2]);
+}
+
+void GshaderToMshader::prepare_pipelines()
+{
+	std::array<VkPipelineShaderStageCreateInfo, 3> shader_stages;
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp]: Maybe use three separate sets of shader_stages for the three pipelines. Then you could have a std::array<..., 2> here, and use static_cast<uint32_t>(shader_stages.size()) down when assigning to the PipelineCreateInfo struct.

File: samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp [hunk 20]
```diff
@@ -0,0 +1,436 @@
...
+	    vkb::initializers::pipeline_dynamic_state_create_info(
+	        dynamic_state_enables.data(),
+	        static_cast<uint32_t>(dynamic_state_enables.size()),
+	        0);
+
+	// Vertex bindings an attributes for model rendering
+
+	// Binding description
+	std::vector<VkVertexInputBindingDescription> vertex_input_bindings = {
+	    vkb::initializers::vertex_input_binding_description(0, sizeof(Vertex), VK_VERTEX_INPUT_RATE_VERTEX)};
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp:181]: Maybe add some comment here, highlighting that you're using `Vertex` here, not `AlignedVertex`

File: samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp [hunk 21]
```diff
@@ -0,0 +1,436 @@
...
+	pipeline_create_info.pColorBlendState    = &color_blend_state;
+	pipeline_create_info.pDynamicState       = &dynamic_state;
+
+	VK_CHECK(vkCreateGraphicsPipelines(get_device().get_handle(), pipeline_cache, 1, &pipeline_create_info, nullptr, &model_pipeline));
+
+	shader_stages[0]                = load_shader("gshader_to_mshader/gshader_to_mshader_base.vert", VK_SHADER_STAGE_VERTEX_BIT);
+	shader_stages[1]                = load_shader("gshader_to_mshader/gshader_to_mshader_base.frag", VK_SHADER_STAGE_FRAGMENT_BIT);
+	shader_stages[2]                = load_shader("gshader_to_mshader/gshader_to_mshader.geom", VK_SHADER_STAGE_GEOMETRY_BIT);
+	pipeline_create_info.stageCount = 3;
+	VK_CHECK(vkCreateGraphicsPipelines(get_device().get_handle(), pipeline_cache, 1, &pipeline_create_info, nullptr, &normal_pipeline));
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp]: Maybe `normal_pipeline` -> `geometry_pipeline` ?

File: samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp [hunk 22]
```diff
@@ -0,0 +1,436 @@
...
+
+	shader_stages[0]                = load_shader("gshader_to_mshader/gshader_to_mshader.mesh", VK_SHADER_STAGE_MESH_BIT_EXT);
+	shader_stages[1]                = load_shader("gshader_to_mshader/gshader_to_mshader_mesh.frag", VK_SHADER_STAGE_FRAGMENT_BIT);
+	pipeline_create_info.stageCount = 2;
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp]: Maybe add some comment that there's no task shader used, and how that maps to how many mesh shaders are spawned on calling `vkCmdDrawMeshTasksEXT`?

File: samples/extensions/gshader_to_mshader/README.md [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+requested_vertex_input_features.meshShader = VK_TRUE;
+```
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/README.md:63]: Maybe add a few word on the pro and cons of both approaches (performance, memory footprint, extensibility, ... ).

File: shaders/gshader_to_mshader/gshader_to_mshader.mesh [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+
+layout (location = 0) out PerVertexData
+{
+  vec4 color;
+} v_out[];
+
+
+void main(void)
+{
+  uint mi = gl_WorkGroupID.x;
```
Comment (asuessenbach) [inline shaders/gshader_to_mshader/gshader_to_mshader.mesh]: Maybe `mi` -> `meshlet_index` ?

**Review by asuessenbach — COMMENTED** [review]
I really like this sample! It's good to have some comparison between geometry shaders and mesh shaders.
There are just a couple of issues.
Besides that, please end all floating point constants in the shaders with an `f`.

File: framework/api_vulkan_sample.h (hunk 3 above)
Comment (Dastin-Wilski) [reply framework/api_vulkan_sample.h]: Thank you for thorough review!
I've tried to address your comments in commit below.

File: framework/gltf_loader.cpp (hunk 4 above)
Comment (Dastin-Wilski) [reply framework/gltf_loader.cpp]: done

File: framework/gltf_loader.cpp (hunk 5 above)
Comment (Dastin-Wilski) [reply framework/gltf_loader.cpp]: Unified these function, does it look acceptable?

File: framework/gltf_loader.cpp (hunk 6 above)
Comment (Dastin-Wilski) [reply framework/gltf_loader.cpp]: fixed

File: framework/gltf_loader.cpp (hunk 7 above)
Comment (Dastin-Wilski) [reply framework/gltf_loader.cpp]: done

File: framework/gltf_loader.cpp (hunk 8 above)
Comment (Dastin-Wilski) [reply framework/gltf_loader.cpp]: added function prepare_meshlets()

File: samples/README.md (hunk 9 above)
Comment (Dastin-Wilski) [reply samples/README.md]: done

File: samples/extensions/gshader_to_mshader/README.md (hunk 10 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/README.md]: done

File: samples/extensions/gshader_to_mshader/README.md (hunk 11 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/README.md]: added more detailed description of a meshlet

File: samples/extensions/gshader_to_mshader/README.md (hunk 12 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/README.md]: done, accounted for unified functions

File: samples/extensions/gshader_to_mshader/README.md (hunk 13 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/README.md]: done

File: samples/extensions/gshader_to_mshader/README.md (hunk 14 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/README.md]: done

File: samples/extensions/gshader_to_mshader/README.md (hunk 15 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/README.md]: done

File: samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp (hunk 16 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp]: fixed

File: samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp (hunk 17 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp]: fixed

File: samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp (hunk 18 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp]: fixed

File: samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp (hunk 19 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp]: done

File: samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp (hunk 20 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp:181]: done

File: samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp (hunk 21 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp]: done

File: samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp (hunk 22 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/gshader_to_mshader.cpp]: done

File: samples/extensions/gshader_to_mshader/README.md (hunk 23 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/README.md:63]: added Pros and Cons headline, it is brief though

File: shaders/gshader_to_mshader/gshader_to_mshader.mesh (hunk 24 above)
Comment (Dastin-Wilski) [reply shaders/gshader_to_mshader/gshader_to_mshader.mesh]: done

File: framework/gltf_loader.cpp [hunk 25]
```diff
@@ -333,6 +333,48 @@ inline void upload_image_to_gpu(CommandBuffer &command_buffer, core::Buffer &sta
...
+				i -= triangle_check;
+				triangle_check = 0;
+			}
+
+			meshlets.push_back(meshlet);
+			meshlet.vertex_count = 0;
+			meshlet.index_count  = 0;
+			vertices.clear();
+		}
+	}
```
Comment (asuessenbach) [inline framework/gltf_loader.cpp:379]: You also need to add the "trailing meshlet" with the last few vertices/indices that did not make it to a full meshlet.

File: framework/gltf_loader.cpp [hunk 26]
```diff
@@ -1043,7 +1085,8 @@ std::unique_ptr<sg::SubMesh> GLTFLoader::load_model(uint32_t index)
 	assert(!gltf_mesh.primitives.empty());
 	auto &gltf_primitive = gltf_mesh.primitives[0];
 
-	std::vector<Vertex> vertex_data;
+	std::vector<Vertex>        vertex_data;
+	std::vector<AlignedVertex> alignedVertex_data;
```
Comment (asuessenbach) [inline framework/gltf_loader.cpp]: Due to naming conventions used here: `alignedVertex_data` -> `aligned_vertex_data` ?

File: samples/extensions/gshader_to_mshader/README.md [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+
+![Sample](./images/visualization_of_normals.png)
+
+## Overview
+
+This sample demonstrates how a mesh shader can be used to achieve the same results as with geometry shader. It contains geometry and mesh shader pipelines visualizing normals in the teapot model.
+
+## Meshlets
+
+To access model vertices from within mesh shader it needs to be stored within an **S**hader **S**torage **B**uffer **O**bject. Indices need to be divided into meshlets and also stored within an SSBO so each work item can work on a single meshlet. Meshlets are created by spliting source geometry.
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/README.md]: Please keep `(SSBO)` at end of first sentence.

File: samples/extensions/gshader_to_mshader/README.md [hunk 28]
```diff
@@ -0,0 +1,70 @@
...
+```C++
+struct Meshlet
+{
+	uint32_t vertices[64];
+	uint32_t indices[126];
+	uint32_t vertex_count;
+	uint32_t index_count;
+};
+```
+By linearly scanning index buffer, meshlet is being filled with index values until 126 indices or 64 unique vertices are reached. New meshlets are added unitl entire model is scanned. `prepare_meshlets()` in gltf_loader.cpp contains exemplary implementation.
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/README.md]: Maybe: "By linearly scanning the indices of the model, Meshlet-structures are created with up to 126 indices or 64 unique vertex indices, whatever is reached first. See `prepare_meshlets()` for an exemplary implementation."

File: samples/extensions/gshader_to_mshader/README.md [hunk 29]
```diff
@@ -0,0 +1,70 @@
...
+
+This sample expands function `load_model()` in the framework, storage_buffer parameter is added (set by defalut to false).
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/README.md]: Maybe: "..., by adding a `bool storage_buffer` parameter, set to `false` by default."

File: samples/extensions/gshader_to_mshader/README.md [hunk 30]
```diff
@@ -0,0 +1,70 @@
...
+By setting storage_buffer to true, model data is read from the file teapot.gltf, vertex position and normals are stored in an SSBO using the `AlignedVertex` structure (because of std430 memory layout). After that, indices are divided into meshlets using the `Meshlet` structure and also stored in SSBOs.
```
Comment (asuessenbach) [inline samples/extensions/gshader_to_mshader/README.md]: Maybe: "In that function model data is read from the file teapot.gltf. By setting `storage_buffer` to `true`, vertex position..."

File: shaders/gshader_to_mshader/gshader_to_mshader.frag [hunk 31]
```diff
@@ -25,8 +25,8 @@ layout (location = 0) out vec4 outFragColor;
 
 void main() 
 {
-	float ambientStrength = 0.2;
-	vec3 ambient = ambientStrength * vec3(1.0, 1.0, 1.0);
+	float ambientStrength_f = 0.2;
```
Comment (asuessenbach) [inline shaders/gshader_to_mshader/gshader_to_mshader.frag]: Ahm... I did not ask to add an f to the variable names, but to the constants: `0.2` - > `0.2f` etc.

File: framework/gltf_loader.cpp (hunk 25 above)
Comment (Dastin-Wilski) [reply framework/gltf_loader.cpp:379]: fixed

File: framework/gltf_loader.cpp (hunk 26 above)
Comment (Dastin-Wilski) [reply framework/gltf_loader.cpp]: done

File: samples/extensions/gshader_to_mshader/README.md (hunk 27 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/README.md]: done

File: samples/extensions/gshader_to_mshader/README.md (hunk 28 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/README.md]: done

File: samples/extensions/gshader_to_mshader/README.md (hunk 29 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/README.md]: done

File: samples/extensions/gshader_to_mshader/README.md (hunk 30 above)
Comment (Dastin-Wilski) [reply samples/extensions/gshader_to_mshader/README.md]: done

File: shaders/gshader_to_mshader/gshader_to_mshader.frag (hunk 31 above)
Comment (Dastin-Wilski) [reply shaders/gshader_to_mshader/gshader_to_mshader.frag]: fixed

**Comment by SaschaWillems** [comment]
This is a great sample!

One thing that we should talk about though is where we actually want to put the ssbo/meshlet creation part. Currently it's part of the framework, so if you want to see how it works you actually how to follow multiple levels of abstraction into the framework. Part of it can be found in the loader, another part (the declaration of a meshlet) on the other hand was added to the api sample base class.

This makes it hard to follow the example, and imo it would be a lot easier to follow this sample if that code was actually part of the sample itself rather than the framework.

**Comment by gpx1000** [comment]
hmm... I was actually thinking one of the nice features of this sample is that it puts the changes into the framework.  This could make it useful for other future samples.  It also could make it easier for developers to see what needs to change from an existing gltf loader.
However, Maybe if we do it in both places it would make sense?  not sure how to make that make sense.

**Comment by SaschaWillems** [comment]
That's true. Just two different point of views ;)

The problem I see with the current approach is that this may make the sample hard to follow. Usually if we show new functionality for the first time, we explicitly do it in the sample itself so you can easily find all things required to get the feature that the sample demonstrates within a single source file (plus header).

So yeah, maybe do it in both places. This new sample is explicit about and new samples that do something with glTF models and mesh shaders can use the framework functions.

**Comment by Dastin-Wilski** [comment]
Maybe making readme more explicit about changes in the framework would be sufficient? It would make easier following the sample. Which structures were added, where and why, which functions where expanded and how. 

On the beginning I was thinking about loading model to SSBO from within the sample but it seemed much easier and less convoluted to fit it into already existing framework. 

In case of having model loading function within the sample I would probably replicate substantial amounts of code from the framework (given my lack of experience with Vulkan and 3D rendering), I thought about having separate, additional files (header and source) within the sample folder dedicated to loading model to SSBO but it seemed like unnecessary boilerplate once I looked into framework  and realized that it provides lion's share of what I need to load model to SSBO.

File: framework/gltf_loader.cpp (hunk 25 above)
Comment (asuessenbach) [reply framework/gltf_loader.cpp:379]: Might be worth now to assert that `triangle_check == 3` when `i == submesh->vertex_indices - 1`

File: shaders/gshader_to_mshader/gshader_to_mshader.frag (hunk 31 above)
Comment (asuessenbach) [reply shaders/gshader_to_mshader/gshader_to_mshader.frag]: Even if it sounds a bit pedantic: I meant _all_ the floating points constants. Like `vec3(1.0f, 1.0f, 1.0f)` etc.

File: framework/gltf_loader.cpp (hunk 25 above)
Comment (Dastin-Wilski) [reply framework/gltf_loader.cpp:379]: done

File: shaders/gshader_to_mshader/gshader_to_mshader.frag (hunk 31 above)
Comment (Dastin-Wilski) [reply shaders/gshader_to_mshader/gshader_to_mshader.frag]: I haven't seen that much in shaders for other samples but done.

**Review by SaschaWillems — APPROVED** [review]
LGTM and works fine on my NV RTX 2060 on Windows 10.

**Review by gpx1000 — APPROVED** [review]
LGTM
